in_toto/verifylib.go

```diff
@@ -0,0 +1,502 @@
+    matched, err := filepath.Match(pattern, name)
+    if err != nil {
+      // The pattern was invalid. We treat it as no match.
+      // TODO: Maybe we should inform the caller at least with a warning?
+      continue
+    }
+    if matched {
+      namesFiltered = append(namesFiltered, name)
+    }
+  }
+  return namesFiltered
+}
+
+
+func VerifyArtifacts(items []interface{}, stepsMetadata map[string]Metablock) error {
+  // Verify artifact rules for each item in the layout
+  for _, itemI := range items {
+    // The layout item (interface) must be a Link or an Inspection
+    // we are only interested in the name and the expected materials and products
+    var itemName string
+    var expected_materials [][]string
+    var expected_products [][]string
+
+    switch item := itemI.(type) {
+      case Step:
+        itemName = item.Name
+        expected_materials = item.ExpectedMaterials
+        expected_products = item.ExpectedProducts
+
+      case Inspection:
+        itemName = item.Name
+        expected_materials = item.ExpectedMaterials
+        expected_products = item.ExpectedProducts
+
+      default: // Something's wrong
+        return fmt.Errorf("VerifyArtifact received an item of invalid type," +
+            " elements of passed slice 'items' must be one of 'Step' or" +
+            " 'Inspection', got: '%s'", reflect.TypeOf(item))
+    }
+    srcLinkMb := stepsMetadata[itemName]
+
+    verificationDataList := []map[string]interface{}{
+      map[string]interface{}{
+        "rules": expected_materials,
+        "artifacts": srcLinkMb.Signed.(Link).Materials,
+      },
+      map[string]interface{}{
+        "rules": expected_products,
+        "artifacts": srcLinkMb.Signed.(Link).Products,
+      },
+    }
+
+    // Process all material rules using the corresponding materials
+    // and all product rules using the corresponding products
+    for _, verificationData := range verificationDataList {
+
+      rules := verificationData["rules"].([][]string)
+      artifacts := verificationData["artifacts"].(map[string]interface{})
+
+      // Create a queue of artifact names (paths)
+      // Each rule only operates on artifacts in that queue
+      // If a rule consumes an artifact (can be applied successfully) it is
+      // removed from the queue.
+      // By applying a DISALLOW rule eventually, verification may return an
+      // error, if the rule matches any artifacts in the queue that should
+      // have been consumed earlier.
+      var queue []string
+      for name, _ := range artifacts {
+        queue = append(queue, name)
+      }
+
+      // Verify rules sequentially
+      for _, rule := range rules {
+        // Parse rule
+        ruleData, err := UnpackRule(rule)
+        if err != nil {
+          return err
+        }
+
+        // Process rules according to rule type
+        // TODO: Currently we only support "MATCH", "ALLOW" and "DISALLOW"
+        switch ruleData["type"] {
+          case "match":
```

Comment:
Comment reads `MATCH`, this checks for lower-case `match`?

Reply:
OK, rules are case-insensitive (db1496bb7ab75a4e72a6191a270c835e67b16ce9), but are the contents of `ruleData`'s entries sure to be lower-cased?

Reply:
I  addressed it in ca976fe. Not sure why the conversation is not marked as outdated.

Reply:
ca976fe addresses the case of words in the comment string; what about the contents of the variable at runtime? Where do they come from? Are the forced to lower case somewhere?

Reply:
Like https://github.com/in-toto/in-toto-golang/pull/1/files#diff-5c3ce1a070ae336476f0f76a5decd5c5R106 , does this upper-case `MATCH` ever end up in `ruleData`?

Reply:
`ruleData` is assigned by calling `rulelib.UnpackRule` above:
 https://github.com/in-toto/in-toto-golang/blob/ca976fee048252062a307af77fe69ac323056425/in_toto/verifylib.go#L138

That function normalizes the case of rule tokens as they appear in the layout.
https://github.com/in-toto/in-toto-golang/blob/ca976fee048252062a307af77fe69ac323056425/in_toto/rulelib.go#L23-L29

Reply:
Thank you very much, I *finally* get it :-)